ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala

```diff
@@ -48,32 +53,130 @@ private class PackageService(reloadPackageStoreIfChanged: PackageService.ReloadP
+              )(state.packageIds)
+            )
+            .map {
+              case Some(diff) =>
+                this.state = this.state.append(diff)
+                logger.info(s"new package IDs loaded: ${diff.keySet.mkString(", ")}")
+                logger.debug(s"loaded diff: $diff")
+              case None => logger.debug(s"new package IDs not found")
+            }
+            .map { res =>
+              lastUpdated = LocalDateTime.now()
+              res
+            }
+            .run
+        try {
+          ongoingPackageUpdateTask = f.map(_ => ().right)
+          // The task future has been updated, now all waiting threads can continue and pick it up.
+          isValid = true
+          updateTaskUpdateSync.release()
+          f.transform { res =>
+            // regardless of whether this was a success or failure
+            // we always want to release the mutex here.
+            isValid = false
+            packageUpdateSync.release()
+            logger.debug("Released mutex after task finished")
+            res
+          }
+        } catch {
+          case NonFatal(t) =>
+            if (updateTaskUpdateSync.availablePermits() == 0)
+              updateTaskUpdateSync.release()
+            packageUpdateSync.release()
+            logger.debug("Released mutex after catching exception")
+            Future.failed(t)
         }
       }
+      // If an update is ongoing
+      // the updateTask var could still contain the old value.
+      // Thus we lock here, to ensure that all threads wait if an
+      // update to that value is potentially incoming.
+      logger.debug("Trying to update packages (tryAcquire)")
+
+      updateTaskUpdateSync.acquire()
+      if (packageUpdateSync.tryAcquire())
+        update()
+      else {
+        updateTaskUpdateSync.release()
+        logger.debug(s"Picking up ongoing update task, is valid? $isValid")
+        ongoingPackageUpdateTask
+      }
     }
 
+    (jwt: Jwt, ec: ExecutionContext, lc: LoggingContextOf[InstanceUUID]) => reload(jwt)(ec, lc)
+  }
+
+  // synchronized via a mutex, so two threads cannot reload it concurrently
+  @inline
+  def reload(jwt: Jwt)(implicit
+      ec: ExecutionContext,
+      lc: LoggingContextOf[InstanceUUID],
+  ) = _reload(jwt, ec, lc)
+
   def packageStore: PackageStore = state.packageStore
 
+  private def withFreshPackagesCheckAndRetry[B](
+      jwt: Jwt
+  )(fn: () => B, checkIfSuccess: B => Boolean)(implicit
+      ec: ExecutionContext,
+      lc: LoggingContextOf[InstanceUUID],
+  ): ET[B] = for {
+    packagesHadToBeFetchedAgain <-
+      if (packagesShouldBeFetchedAgain) EitherT(reload(jwt)).map(_ => true)
+      else EitherT.pure(false): ET[Boolean]
+    _ = logger.trace(s"Did we have to reload the packages: $packagesHadToBeFetchedAgain")
+    result <- EitherT.pure(fn()): ET[B]
+    _ = logger.trace(s"Result: $result")
+    finalResult <-
+      if (!checkIfSuccess(result) && !packagesHadToBeFetchedAgain) {
```

Comment:
I can adjust it to be that fine grained. In the end my approach was just the easiest initial solution to me.

Reply:
@cocreature whats the reason behind the refresh on point 4 ? i.e if we have the package (but no package id) what do we expect to get from the refresh ?

Reply:
the json API errors out if the module name is not unique across modules. If we never refresh on 4 we will keep selecting a package even if it is no longer unique.

Reply:
> 1. package id and we have the package

Do you actually mean here:

`package id provided which is in our cache and a template id could be found`

Reply:
no if you know the package and you don’t find it, you also don’t need to refresh. packages don’t mutate so a refresh isn’t going to result in a successful lookup.

Reply:
ok for point 4 however we need to assume that the search was successful, otherwise how should I know that we have the package if no package id is provided :D?

Reply:
For 3 you do, for 4 you don’t. If the refresh fails you just resolve to the package you already know. The refresh in 4 is only to turn a successful resolve into an ambiguity error. We want to make sure that this happens at some point but I don’t think it should fail any individual request.